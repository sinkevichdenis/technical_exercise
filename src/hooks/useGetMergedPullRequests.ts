import { useEffect, useReducer } from 'react';

import { apiReducer, initialState, StateType } from 'api/apiReducer';
import { ResultType } from 'api/getMergedPullRequests';

import { fetchData } from './utils/fetchData';


export const useGetMergedPullRequests = (): StateType<ResultType> & { refetch: () => void } => {
  const [data, dispatch] = useReducer(apiReducer<ResultType>, initialState);

  useEffect(() => {
    fetchData(dispatch);
  }, []);

  return {
    ...data,
    refetch: () => fetchData(dispatch),
  };
};
