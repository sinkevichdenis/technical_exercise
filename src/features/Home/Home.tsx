import React from 'react';
import { ScrollView, StyleSheet, Text, View } from 'react-native';
import { SafeAreaView } from 'react-native-safe-area-context';

import { Button } from 'components/Button';
import { ErrorMessageArea } from 'components/ErrorMessageArea';
import { useGetMergedPullRequests } from 'hooks/useGetMergedPullRequests';
import { Colors } from 'styles/colors';
import { DEFAULT_MARGIN } from 'styles/consts';

import { LoadingSpinner } from '../../components/LoadingSpinner';


export const Home: React.FC = React.memo(() => {
  const { data, loading, error, refetch } = useGetMergedPullRequests();
  return (
    <SafeAreaView style={ styles.safeArea }>
      <ScrollView
        style={ styles.container }
        bounces={ false }
      >
        { loading ? <LoadingSpinner containerStyle={ styles.marginTop }/> : (
          <View
            style={ styles.contentWrapper }
            testID='home-content'
          >
            <Text style={ styles.text }>
              { data?.length }
            </Text>
            { error && (
              <ErrorMessageArea
                text={ `${error.status}:\n${error.message}` }
                onPress={ refetch }
              />
            ) }
            { !error && !!data.length && (
              <Button
                onPress={ refetch }
                text='Update'
                containerStyle={ styles.marginTop }
                testID='update-btn'
              />
            ) }
          </View>
        ) }
      </ScrollView>
    </SafeAreaView>
  );
});

const styles = StyleSheet.create({
  safeArea: {
    flex: 1,
    backgroundColor: Colors.WHITE,
  },
  container: {
    padding: DEFAULT_MARGIN,
  },
  contentWrapper: {
    marginTop: DEFAULT_MARGIN * 3,
  },
  text: {
    color: Colors.TEXT_BLACK,
  },
  marginTop: {
    marginTop: DEFAULT_MARGIN,
  },
});
