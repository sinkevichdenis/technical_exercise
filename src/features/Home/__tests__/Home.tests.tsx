import { render } from '@testing-library/react-native';
import React from 'react';


import { mockOctokitFilteredResponse } from '__mocks__/octokit.mock';
import { useGetMergedPullRequests } from 'hooks/useGetMergedPullRequests';

import { Home } from '../Home';


const mockInitialState = { data: [], error: null, loading: false };
const mockSuccessState = { data: mockOctokitFilteredResponse, error: null, loading: false };
const mockLoadingState = { data: mockOctokitFilteredResponse, error: null, loading: true };
const mockErrorState = { data: [], error: { message: 'error' }, loading: false };

const errorId = 'error-message-area';
const buttonId = 'update-btn';
const loaderId = 'spinner';


jest.mock('hooks/useGetMergedPullRequests');

describe('Home component', () => {
  const componentRender = () => render(<Home />);

  it('handle initial state without data', () => {
    (useGetMergedPullRequests as jest.Mock).mockReturnValue(mockInitialState);
    const { toJSON, queryByTestId } = componentRender();
    expect(toJSON()).toMatchSnapshot();
    expect(queryByTestId(errorId)).toBeNull();
    expect(queryByTestId(buttonId)).toBeNull();
    expect(queryByTestId(loaderId)).toBeNull();
  });

  it('handle successful state with data', () => {
    (useGetMergedPullRequests as jest.Mock).mockReturnValue(mockSuccessState);
    const { toJSON, queryByTestId } = componentRender();
    expect(toJSON()).toMatchSnapshot();
    expect(queryByTestId(errorId)).toBeNull();
    expect(queryByTestId(buttonId)).toBeTruthy();
    expect(queryByTestId(loaderId)).toBeNull();
  });

  it('handle loading state', () => {
    (useGetMergedPullRequests as jest.Mock).mockReturnValue(mockLoadingState);
    const { queryByTestId } = componentRender();
    expect(queryByTestId('home-content')).toBeNull();
    expect(queryByTestId(loaderId)).toBeTruthy();
  });

  it('handle error state', () => {
    (useGetMergedPullRequests as jest.Mock).mockReturnValue(mockErrorState);
    const { toJSON, queryByTestId } = componentRender();
    expect(toJSON()).toMatchSnapshot();
    expect(queryByTestId(errorId)).toBeTruthy();
    expect(queryByTestId(buttonId)).toBeNull();
    expect(queryByTestId(loaderId)).toBeNull();
  });
});
