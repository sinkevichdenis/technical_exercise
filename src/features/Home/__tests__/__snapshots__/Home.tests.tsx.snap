// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component handle error state 1`] = `
<RNCSafeAreaView
  edges={
    {
      "bottom": "additive",
      "left": "additive",
      "right": "additive",
      "top": "additive",
    }
  }
  style={
    {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    bounces={false}
    style={
      {
        "padding": 16,
      }
    }
  >
    <View>
      <View
        style={
          {
            "marginTop": 48,
          }
        }
        testID="home-content"
      >
        <Text
          style={
            {
              "color": "#1D2330",
            }
          }
        >
          This is an example with mock data. Please try refetch a real data.
        </Text>
        <View
          style={
            {
              "marginVertical": 8,
            }
          }
          testID="histogram"
        />
        <View
          style={
            [
              {
                "borderColor": "#FF3B30",
                "borderRadius": 12,
                "borderWidth": 1,
                "padding": 16,
              },
              undefined,
            ]
          }
          testID="error-message-area"
        >
          <Text
            style={
              {
                "color": "#FF3B30",
                "fontSize": 16,
                "lineHeight": 20,
              }
            }
          >
            undefined:
error
          </Text>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={false}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#1D2330",
                "borderRadius": 12,
                "borderWidth": 1,
                "justifyContent": "center",
                "marginTop": 16,
                "opacity": 1,
                "paddingHorizontal": 16,
                "paddingVertical": 8,
              }
            }
            testID="error-area-btn"
          >
            <Text
              style={
                [
                  {
                    "color": "#FFFFFF",
                    "fontSize": 16,
                    "lineHeight": 20,
                  },
                  undefined,
                ]
              }
            >
              Try again
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RNCSafeAreaView>
`;

exports[`Home component handle initial state without data 1`] = `
<RNCSafeAreaView
  edges={
    {
      "bottom": "additive",
      "left": "additive",
      "right": "additive",
      "top": "additive",
    }
  }
  style={
    {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    bounces={false}
    style={
      {
        "padding": 16,
      }
    }
  >
    <View>
      <View
        style={
          {
            "marginTop": 48,
          }
        }
        testID="home-content"
      >
        <Text
          style={
            {
              "color": "#1D2330",
            }
          }
        >
          This is an example with mock data. Please try refetch a real data.
        </Text>
        <View
          style={
            {
              "marginVertical": 8,
            }
          }
          testID="histogram"
        />
      </View>
    </View>
  </RCTScrollView>
</RNCSafeAreaView>
`;

exports[`Home component handle successful state with data 1`] = `
<RNCSafeAreaView
  edges={
    {
      "bottom": "additive",
      "left": "additive",
      "right": "additive",
      "top": "additive",
    }
  }
  style={
    {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    bounces={false}
    style={
      {
        "padding": 16,
      }
    }
  >
    <View>
      <View
        style={
          {
            "marginTop": 48,
          }
        }
        testID="home-content"
      >
        <Text
          style={
            {
              "color": "#1D2330",
            }
          }
        >
          This is a real data!!
        </Text>
        <View
          style={
            {
              "marginVertical": 8,
            }
          }
          testID="histogram"
        />
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={false}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#1D2330",
              "borderRadius": 12,
              "borderWidth": 1,
              "justifyContent": "center",
              "marginTop": 16,
              "opacity": 1,
              "paddingHorizontal": 16,
              "paddingVertical": 8,
            }
          }
          testID="update-btn"
        >
          <Text
            style={
              [
                {
                  "color": "#FFFFFF",
                  "fontSize": 16,
                  "lineHeight": 20,
                },
                undefined,
              ]
            }
          >
            Update
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RNCSafeAreaView>
`;
